rclrs/src/node/client.rs

```diff
@@ -0,0 +1,257 @@
+use crate::node::client::oneshot::Canceled;
+use futures::channel::oneshot;
+use std::borrow::Borrow;
+use std::boxed::Box;
+use std::collections::HashMap;
+use std::ffi::CString;
+use std::sync::atomic::{AtomicI64, Ordering};
+use std::sync::Arc;
+
+use crate::error::{RclReturnCode, ToResult};
+use crate::MessageCow;
+use crate::Node;
+use crate::{rcl_bindings::*, RclrsError};
+
+use parking_lot::{Mutex, MutexGuard};
+use rosidl_runtime_rs::Message;
+
+// SAFETY: The functions accessing this type, including drop(), shouldn't care about the thread
+// they are running in. Therefore, this type can be safely sent to another thread.
+unsafe impl Send for rcl_client_t {}
+
+/// Internal struct used by clients.
+pub struct ClientHandle {
+    rcl_client_mtx: Mutex<rcl_client_t>,
+    rcl_node_mtx: Arc<Mutex<rcl_node_t>>,
+}
+
+impl ClientHandle {
+    pub(crate) fn lock(&self) -> MutexGuard<rcl_client_t> {
+        self.rcl_client_mtx.lock()
+    }
+}
+
+impl Drop for ClientHandle {
+    fn drop(&mut self) {
+        let handle = self.rcl_client_mtx.get_mut();
+        let rcl_node_mtx = &mut *self.rcl_node_mtx.lock();
+        unsafe {
+            rcl_client_fini(handle as *mut _, rcl_node_mtx as *mut _);
+        }
+    }
+}
+
+impl From<Canceled> for RclrsError {
+    fn from(_: Canceled) -> Self {
+        RclrsError::RclError {
+            code: RclReturnCode::Error,
+            msg: None,
```

Comment:
Should we actually have a message here? Something to let the user know that the request was cancelled when they're doing something like inspecting the bag files?

Would such a thing be useful?

Reply:
Am I correct to assume that an occurrence of `Canceled` would basically mean we have a logic error in our code?

Reply:
I don't _think_ so... I think you can get `Canceled` requests in a variety of situations, though I'm not 100% positive how common they are. So it's not necessarily a logic error, just a way to stop the service from doing extra work if it isn't needed.

Reply:
Do you mean using `unwrap()` after the future is inserted into the hashmap? If that's the case, yes `unwrap()` should be safe there, otherwise please elaborate.

Reply:
Exactly, in the last statement of `call_async()`, I'd write `rx.await.unwrap()`.